### PR TITLE
[SYSTEMDS-2622] Added LARS Optimizer

### PR DESCRIPTION
**Pull request by JonahBalshai**
## Summary

This PR implements the [**Layer-wise Adaptive Rate Scaling (LARS)**](https://arxiv.org/pdf/1708.03888) optimizer in SystemDS. LARS is designed to improve convergence for large-batch training by scaling the learning rate of each layer based on the norm of its weights and gradients. This is especially useful in deep learning applications where standard optimizers like SGD struggle to scale efficiently.

## Changes

- Added `lars.dml` and `lars_util.dml` to `scripts/nn/optim/`
- Modified all instances of ResNet in  `scripts/nn/networks/` to allow for optimization using LARS
- Added `alexnet.dml`  to  `scripts/nn/networks/`
- Added local response normalization layer `lrn.dml` to  `scripts/nn/layers/`
- Extended example scripts with LARS examples in  `scripts/nn/examples/`

## Performance
We tested LARS on the MNIST dataset with varying batch sizes against other optimizers. In the plots below you can see how our implementation of LARS shapes up against other optimizers compared to batch sizes.

PLOTS HERE. With some explanations

## Challenges and Issues
- Due to the limited data loading capabilities of systemds and our own hardware we often ran into memory issues when loading larger batches. Especially with datasets containing larger images like ImageNet.

We're happy to receive any feedback or suggestions to improve our implementation.

### CODE REVIEW

File: src/test/java/org/apache/sysds/test/functions/mlcontext/MLContextTest.java [hunk 1]
```diff
@@ -1965,28 +1964,4 @@ public void testNNImport() {
 			.getScalarObject("R").getDoubleValue();
 		Assert.assertEquals(1000, ret, 1e-20);
 	}
-
```
Comment (christinadionysio) [inline src/test/java/org/apache/sysds/test/functions/mlcontext/MLContextTest.java]: Why were these tests removed?

**Review by christinadionysio — CHANGES_REQUESTED** [review]
Please add the missing license headers to all new files you create. 
Could you please also have a look into the failing ResNet test?

**Comment by JonahBalshai** [comment]
> Please add the missing license headers to all new files you create. Could you please also have a look into the failing ResNet test?

Hi Christina, thanks for the review! Could you be more specific in regards to which resnet test fails? It's easier for us since starting up a test takes quite some time. Thanks :)

File: src/test/java/org/apache/sysds/test/functions/mlcontext/MLContextTest.java (hunk 1 above)
Comment (JonahBalshai) [reply src/test/java/org/apache/sysds/test/functions/mlcontext/MLContextTest.java]: These seem to have been removed by accident. We put them back in.

**Comment by christinadionysio** [comment]
> > Please add the missing license headers to all new files you create. Could you please also have a look into the failing ResNet test?
> 
> Hi Christina, thanks for the review! Could you be more specific in regards to which resnet test fails? It's easier for us since starting up a test takes quite some time. Thanks :)

The particular test that fails is this one: org.apache.sysds.test.applications.nn.NNComponentTest.resnet 
You can have a look at the failing Java Test github action on this PR to see the details.

**Comment by Javiermateor** [comment]
Hi @christinadionysio ,

 I  ran the failing federated tests locally to rule out any issues with my changes. Here are the results:

*   `mvn test -Dtest=...FederatedFullCumulativeTest` -> **BUILD SUCCESS** (16 tests passed)
*   `mvn test -Dtest=...FederatedFullAggregateTest` -> **BUILD SUCCESS** (22 tests passed)

**Both of the test classes that are failing in the CI pipeline pass perfectly on my local machine.** The failures may be due to a flaky test in the CI environment and are not caused by my changes in our `nn` module or in our implementation for the PR.

For completeness, the issue in the CI appears to be related to `FED°fedinit` instructions, which points towards other error sources, maybe environmental factors like networking or process timing.

This pattern strongly suggests that these are flaky tests and the failures are due to the CI environment rather than a bug introduced by my PR.

Could a maintainer please take a look or advise on how to proceed?

**Comment by christinadionysio** [comment]
> Hi @christinadionysio ,
> 
> I ran the failing federated tests locally to rule out any issues with my changes. Here are the results:
> 
> * `mvn test -Dtest=...FederatedFullCumulativeTest` -> **BUILD SUCCESS** (16 tests passed)
> * `mvn test -Dtest=...FederatedFullAggregateTest` -> **BUILD SUCCESS** (22 tests passed)
> 
> **Both of the test classes that are failing in the CI pipeline pass perfectly on my local machine.** The failures may be due to a flaky test in the CI environment and are not caused by my changes in our `nn` module or in our implementation for the PR.
> 
> For completeness, the issue in the CI appears to be related to `FED°fedinit` instructions, which points towards other error sources, maybe environmental factors like networking or process timing.
> 
> This pattern strongly suggests that these are flaky tests and the failures are due to the CI environment rather than a bug introduced by my PR.
> 
> Could a maintainer please take a look or advise on how to proceed?

Hi @Javiermateor!

Thank you for your comment. 
We know about this issue and are already investigating it, so the tests are not failing because of your implementation.

**Comment by e-strauss** [comment]
@Javiermateor @christinadionysio Hi, I will rerun the CI tests. Then they should pass. Thanks

**Comment by JonahBalshai** [comment]
Hey! Sorry for the late response. We honestly thought the branch would be auto-merged after approval 😄. I just pulled all the upstream commits into the branch which restarted the pipeline. Should we redo the pipeline or just merge the PR?
